fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/directory/LuceneSerializer.java

```diff
@@ -0,0 +1,189 @@
+        if (data == null) {
+            return null;
+        }
+        final EncodingState state = new EncodingState();
+        if (compress) {
+            data = compress(data, state);
+        }
+        if (encrypt) {
+            data = encrypt(data, state);
+        }
+        final byte[] encoded = new byte[data.length + 1];
+        int code = 0;
+        if (state.isCompressed()) {
+            code = code | ENCODING_COMPRESSED;
+        }
+        if (state.isEncrypted()) {
+            code = code | ENCODING_ENCRYPTED;
+        }
+        encoded[0] = (byte) code;
+        System.arraycopy(data, 0, encoded, 1, data.length);
+        LOGGER.trace("Encoded data for resourceDescription={}, compressionSupposed={}, encryptionSupposed={}, compressedEventually={}, encryptedEventually={}, originalDataSize={}, encodedDataSize={}",
+                resourceDescription, compress, encrypt, state.isCompressed(), state.isEncrypted(), data.length, encoded.length);
+        return encoded;
+    }
+
+    @Nullable
+    public static byte[] decode(@Nonnull String resourceDescription, @Nullable byte[] data) {
+        if (data == null) {
+            return null;
+        }
+        if (data.length < 2) {
+            throw new RecordCoreException("Invalid data")
+                    .addLogInfo(LogMessageKeys.DATA_VALUE, data);
+        }
+
+        final byte encoding = data[0];
+        byte[] decoded = new byte[data.length - 1];
+        System.arraycopy(data, 1, decoded, 0, data.length - 1);
+
+        final boolean encrypted = (encoding & ENCODING_ENCRYPTED) == ENCODING_ENCRYPTED;
+        final boolean compressed = (encoding & ENCODING_COMPRESSED) == ENCODING_COMPRESSED;
+        if (encrypted) {
+            decoded = decrypt(decoded);
+        }
+        if (compressed) {
+            decoded = decompress(decoded);
+        }
+        LOGGER.trace("Decoded data for resourceDescription={}, compressed={}, encrypted={}, encodedDataSize={}, decodedDataSize={}",
+                resourceDescription, compressed, encrypted, data.length, decoded.length);
+        return decoded;
+    }
+
+    @Nonnull
+    private static byte[] compress(@Nonnull byte[] data, EncodingState state) {
+        try (Compressor compressor = CompressionMode.HIGH_COMPRESSION.newCompressor()) {
+            final ByteBuffersDataOutput output = new ByteBuffersDataOutput();
+            output.writeVInt(data.length);
+            compressor.compress(data, 0, data.length, output);
+            final byte[] compressedBytes = output.toArrayCopy();
+
+            // Compress only if it helps to shorten the bytes
+            if (compressedBytes.length + 1 < data.length) {
+                state.setCompressed(true);
+                final byte[] compressedData = new byte[compressedBytes.length + 1];
+                compressedData[0] = (byte) 0; // The version 0 representing HIGH_COMPRESSION, this can be extended to support more compression modes in future
```

Comment:
Isn't that `COMPRESSION_VERSION_FOR_HIGH_COMPRESSION`? Saving a comment and a cast?

Reply:
Fixed.